src/compiler/checker.ts

```diff
@@ -12859,8 +12859,12 @@ namespace ts {
             const lastParamVariadicType = firstDefined(lastParamTags, p =>
                 p.typeExpression && isJSDocVariadicType(p.typeExpression.type) ? p.typeExpression.type : undefined);
 
-            const syntheticArgsSymbol = createSymbol(SymbolFlags.Variable, "args" as __String, CheckFlags.RestParameter);
-            syntheticArgsSymbol.type = lastParamVariadicType ? createArrayType(getTypeFromTypeNode(lastParamVariadicType.type)) : anyArrayType;
+            // By using a `DeferredType` symbol, we allow the type of this rest arg to be overriden by contextual type assignment so long as its type hasn't been
+            // cached by `getTypeOfSymbol` yet.
+            const syntheticArgsSymbol = createSymbol(SymbolFlags.Variable, "args" as __String, CheckFlags.RestParameter | CheckFlags.DeferredType);
+            syntheticArgsSymbol.deferralParent = neverType;

```

Comment:
Since it's a single-element deferral, the `.parentType` doesn't actually matter; we'll just promote the deferred member to the full cached type upon request (via single-element fastpaths in `getUnionType` or `getIntersectionType`).